hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -310,7 +310,15 @@ def __init__(self, resp):
 
     async def read(self, n: int = -1) -> bytes:
         assert not self._closed
-        return await self._content.read(n)
+        data = bytearray()
+        while n == -1 or n > 0:
+            b = await self._content.read(n)
+            if not b:
+                break
+            data.extend(b)
+            if n != -1:
+                n -= len(b)
+        return bytes(data)
```

Comment:
Hmm. If `not b`, and `n > 0` then we must raise an UnexpectedEOF, right? 

I find it easier to convince myself I'm right if I write it in two pieces like this:

```python3
data = bytearray()

if n == -1:
    b = await self._content.read(n)
    data.extend(b)
    while b:
        b = await self._content.read(n)
        data.extend(b)
    return bytes(data)

while n > 0:
    b = await self._content.read(n)
    if not b:
        raise UnexpectedEOF()
    data.extend(b)
    n -= len(b)
assert n == 0
return bytes(data)

```

Reply:
Actually, the python interface just gives you at most n bytes requested. If you request 5 bytes and there's only 3, then you just get 3 bytes with no error.

Reply:
Something seems off to me. The contract of Python's `read(size)` is:
> Read up to size bytes from the object and return them.

The contract of the `read(n)` in this PR is:
> Read exactly size bytes from the object unless we encounter EOF in which case read fewer than size bytes.


Who uses `GetObjectStream`? Why do they want the aforementioned contract? I'd think think they either want the contract of Python's read or the "read exactly n bytes or raise" contract.

Reply:
`GetObjectStream` is returned by `StorageClient.get_object` which is returned by `GoogleStorageAsyncFS.read` and `GoogleStorageAsyncFS.open_from`. So when we read a file, we call `read` on the `GetObjectStream`. The LocalAsyncFS reads from a file handle. This returns all the bytes. For example, we write code like this when using the Python `open` function.

```python3
with open('file', 'r') as f:
     data = f.read()
```

So `AsyncFS.read()` can not be used like this if we don't return all bytes when reading from the stream.

```python3
    async def read(self, url: str) -> bytes:
        async with await self.open(url) as f:
            return await f.read()
```

It would need to be like this:

```python3
    async def read(self, url: str) -> bytes:
        async with await self.open(url) as f:
            data = bytearray()
            b = f.read()
            data.extend(b)
            while b:
                 b = await f.read()
                 data.extend(b)
```